src/app/data-model-provider/MetadataList.cpp

```diff
@@ -25,106 +25,158 @@ namespace app {
-        other.mIsImmutable  = true;
+        Platform::MemoryFree(mBuffer);
     }
-    return *this;
 }
 
-const void * GenericMetadataList::operator[](size_t index) const
+GenericAppendOnlyBuffer::GenericAppendOnlyBuffer(GenericAppendOnlyBuffer && other) : mElementSize(other.mElementSize)
 {
-    VerifyOrDie(index < mElementCount);
-    return mBuffer + index * mElementSize;
+    // take over the data
+    mBuffer            = other.mBuffer;
+    mElementCount      = other.mElementCount;
+    mCapacity          = other.mCapacity;
+    mBufferIsAllocated = other.mBufferIsAllocated;
+
+    // clear other
+    other.mBuffer            = nullptr;
+    other.mElementCount      = 0;
+    other.mCapacity          = 0;
+    other.mBufferIsAllocated = false;
 }
 
-CHIP_ERROR GenericMetadataList::Reserve(size_t numElements)
+GenericAppendOnlyBuffer & GenericAppendOnlyBuffer::operator=(GenericAppendOnlyBuffer && other)
 {
-    VerifyOrReturnError(!mIsImmutable, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mElementCount == 0, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrDie(mElementSize == other.mElementSize);
+
+    if (mBufferIsAllocated && (mBuffer != nullptr))
+    {
+        Platform::Impl::PlatformMemoryManagement::MemoryFree(mBuffer);
+    }
+
+    // take over the data
+    mBuffer            = other.mBuffer;
+    mElementCount      = other.mElementCount;
+    mCapacity          = other.mCapacity;
+    mBufferIsAllocated = other.mBufferIsAllocated;
+
+    // clear other
+    other.mBuffer            = nullptr;
+    other.mElementCount      = 0;
+    other.mCapacity          = 0;
+    other.mBufferIsAllocated = false;
 
-    if ((mBuffer != nullptr) && mAllocated)
+    return *this;
+}
+
+CHIP_ERROR GenericAppendOnlyBuffer::EnsureAppendCapacity(size_t numElements)
+{
+    if (mCapacity >= mElementCount + numElements)
     {
-        chip::Platform::MemoryFree(mBuffer);
+        // Sufficient capacity already exists
+        return CHIP_NO_ERROR;
     }
 
-    mBuffer = static_cast<uint8_t *>(chip::Platform::MemoryCalloc(numElements, mElementSize));
+    if (mBuffer == nullptr)
+    {
+        mBuffer            = static_cast<uint8_t *>(Platform::MemoryCalloc(numElements, mElementSize));
+        mCapacity          = numElements;
+        mBufferIsAllocated = true;
+        return mBuffer != nullptr ? CHIP_NO_ERROR : CHIP_ERROR_NO_MEMORY;
+    }
 
-    VerifyOrReturnError(mBuffer != nullptr, CHIP_ERROR_NO_MEMORY);
+    // we already have the data in buffer. we have two choices:
+    //   - allocated buffer needs to be extended
+    //   - re-used const buffer neext to be copied over
+    if (mBufferIsAllocated)
+    {
+        auto new_buffer = static_cast<uint8_t *>(Platform::MemoryRealloc(mBuffer, (mElementCount + numElements) * mElementSize));
+        VerifyOrReturnError(new_buffer != nullptr, CHIP_ERROR_NO_MEMORY);
+        mBuffer = new_buffer;
+    }
+    else
+    {
+        // this is NOT an allocated buffer, but it should become one
```

Comment:
Is this comment true?

Reply:
Yes: the above condition is `mBufferIsAllocated` so else means it is not allocated (but because it is non-empty it has to be allocated to append things to it)